src/neo/SmartContract/Native/NativeContract.cs

```diff
@@ -14,32 +14,36 @@ namespace Neo.SmartContract.Native
     public abstract class NativeContract
     {
         private static readonly List<NativeContract> contractsList = new List<NativeContract>();
-        private static readonly Dictionary<string, NativeContract> contractsNameDictionary = new Dictionary<string, NativeContract>();
+        private static readonly Dictionary<int, NativeContract> contractsIdDictionary = new Dictionary<int, NativeContract>();
         private static readonly Dictionary<UInt160, NativeContract> contractsHashDictionary = new Dictionary<UInt160, NativeContract>();
         private readonly Dictionary<string, ContractMethodMetadata> methods = new Dictionary<string, ContractMethodMetadata>();
+        private static int id_counter = 0;
 
-        public static IReadOnlyCollection<NativeContract> Contracts { get; } = contractsList;
+        #region Named Native Contracts
         public static ContractManagement ContractManagement { get; } = new ContractManagement();
```

Comment:
Maybe we can init this contracts inside the static constructor in order to avoid that the compiler optimizer play with us.

Reply:
Usually only partial classes have non-deterministic problems.

Reply:
I think that it's safer because this part is uncontrolled by us.

Reply:
But I don't want to modify every native contract each time we create a new one.

Reply:
Maybe if we use this code, we can avoid it?

```csharp
[MethodImpl(MethodImplOptions.NoOptimization)]
static NativeContract() { }
```

Reply:
It won't be any problem without this code.